src/scportrait/pipeline/extraction.py

```diff
@@ -524,206 +509,145 @@ def _extract_classes(self, px_center, arg, return_failed_ids=False):
-            # visualize a random cell for every 100 contained in the dataset
-            n_cells = 100
-            n_cells_to_visualize = len(keep_index) // n_cells
-
-            rng = np.random.default_rng()
-            random_indexes = rng.choice(keep_index, n_cells_to_visualize, replace=False)
-
-            for index in random_indexes:
-                stack = _tmp_single_cell_data[index]
+    def _create_output_files(self):
+        single_cell_index_shape = (self.num_classes, 2)
+        single_cell_data_shape = (
+            self.num_classes,
+            (self.n_masks + self.n_image_channels),
+            self.image_size,
+            self.image_size,
+        )
 
-                fig, axs = plt.subplots(1, stack.shape[0], figsize=(2 * stack.shape[0], 2))
-                for i, img in enumerate(stack):
-                    axs[i].imshow(img, vmin=0, vmax=1)
-                    axs[i].axis("off")
-                fig.tight_layout()
-                plt.show(fig)
+        # generate container for single_cell_data
+        self._tmp_single_cell_data_path = create_empty_mmap(
+            shape=single_cell_data_shape,
+            dtype=np.float16,
+            tmp_dir_abs_path=self._tmp_dir_path,
+        )
 
-        self.log("Transferring extracted single cells to .hdf5")
+        # generate container for single_cell_index
+        fixed_length = 200  # this is the maximum length of string that can be stored in the mmap array for the cell_ids
+        dt = np.dtype(f"S{fixed_length}")
+        self._tmp_single_cell_index_path = create_empty_mmap(
+            shape=single_cell_index_shape,
+            dtype=dt,
+            tmp_dir_abs_path=self._tmp_dir_path,
+        )
 
         # create name for output file
         self.output_path = os.path.join(self.extraction_data_directory, self.DEFAULT_EXTRACTION_FILE)
 
         with h5py.File(self.output_path, "w") as hf:
             hf.create_dataset(
                 "single_cell_index",
-                data=list(zip(list(range(len(cell_ids))), cell_ids, strict=False)),
+                shape=single_cell_index_shape,
                 dtype=self.DEFAULT_SEGMENTATION_DTYPE,
             )  # increase to 64 bit otherwise information may become truncated
 
-            self.log("single-cell index created.")
-            del cell_ids
-            # self._clear_cache(vars_to_delete=[cell_ids]) # this is not working as expected so we will just delete the variable directly
+            self.log("Container for single-cell index created.")
 
-            _, c, x, y = _tmp_single_cell_data.shape
-            single_cell_data = hf.create_dataset(
+            hf.create_dataset(
                 "single_cell_data",
-                shape=(len(keep_index), c, x, y),
+                shape=single_cell_data_shape,
                 chunks=(1, 1, self.image_size, self.image_size),
                 compression=self.compression_type,
                 dtype=np.float16,
```

Comment:
why this dtype?

Reply:
could you please clarify this? Not sure what you mean.